pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -0,0 +1,353 @@
+	Email             string   `json:"email"`
+	EmailVerified     bool     `json:"email_verified"`
+	Groups            []string `json:"groups"`
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, tokenMGR *tokens.Manager) common.AuthProvider {
+	return &oidcProvider{
+		ctx:         ctx,
+		authConfigs: mgmtCtx.Management.AuthConfigs(""),
+		secrets:     mgmtCtx.Core.Secrets(""),
+		userMGR:     userMGR,
+		tokenMGR:    tokenMGR,
+	}
+}
+
+func (o *oidcProvider) GetName() string {
+	return Name
+}
+
+func (o *oidcProvider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = o.actionHandler
+	schema.Formatter = o.formatter
+}
+
+func (o *oidcProvider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	login, ok := input.(*v32.OIDCLogin)
+	if !ok {
+		return v3.Principal{}, nil, "", fmt.Errorf("unexpected input type")
+	}
+	return o.LoginUser(ctx, login, nil)
+}
+
+func (o *oidcProvider) LoginUser(ctx context.Context, oauthLoginInfo *v32.OIDCLogin, config *v32.OIDCConfig) (v3.Principal, []v3.Principal, string, error) {
+	var userPrincipal v3.Principal
+	var groupPrincipal []v3.Principal
+	var claimInfo ClaimInfo
+	var err error
+
+	if config == nil {
+		config, err = o.getOIDCConfig()
+		if err != nil {
+			return userPrincipal, nil, "", err
+		}
+	}
+
+	provider, err := oidc.NewProvider(ctx, config.Issuer)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+
+	oauthConfig := oauth2.Config{
+		ClientID:     config.ClientID,
+		ClientSecret: config.ClientSecret,
+		Endpoint:     provider.Endpoint(),
+		RedirectURL:  config.RancherURL,
+		Scopes:       []string{config.Scopes},
+	}
+
+	logrus.Debugf("[generic oidc] loginuser: using code to get oauth token")
+	oauth2Token, err := oauthConfig.Exchange(ctx, oauthLoginInfo.Code)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+
+	userInfo, err := provider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token))
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+
+	userInfo.Claims(&claimInfo)
+	userPrincipal = o.userToPrincipal(userInfo, claimInfo)
+
+	for _, group := range claimInfo.Groups {
+		groupPrincipal = append(groupPrincipal, o.groupToPrincipal(group))
+	}
+	logrus.Debugf("[generic oidc] loginuser: Checking user's access to Rancher")
+	allowed, err := o.userMGR.CheckAccess(config.AccessMode, config.AllowedPrincipalIDs, userPrincipal.Name, groupPrincipal)
+	if err != nil {
+		return userPrincipal, groupPrincipal, "", err
+	}
+	if !allowed {
+		return userPrincipal, groupPrincipal, "", httperror.NewAPIError(httperror.Unauthorized, "unauthorized")
+	}
+	return userPrincipal, groupPrincipal, oauth2Token.AccessToken, nil
+}
+
+func (o *oidcProvider) SearchPrincipals(searchValue, principalType string, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	if principalType == "" {
+		principalType = UserType
+	}
+
```

Comment:
Again, this is how SAML is set up - since we don't have the ability to query all the users in the user management system at this time, then all we can return is a non search result. We could always opt for a "no implemented" error instead if that would be preferred.